Beta rc

Connoropolous opened the pull request:
This is:
- [x] version bumps
- [x] introduces some necessary changes to JS libs because of the need to be able to handle the authorization of zome calls  (https://github.com/h-REA/hREA/pull/376/files?file-filters%5B%5D=.js&file-filters%5B%5D=.ts&show-viewed-files=true)  which must be done via the admin API. I can imagine making changes to this in the future based on learning about real world security contexts
- [x] update run-dev-conductor.sh @Connoropolous 
- [x] getting the GraphQL explorer working again @Connoropolous 
- [x] fix indexing issue with failing tests

Tests that are failing seem to be related to read_all_entry_hashes and the time index. It seems to be returning a longer list of results than it should.
https://github.com/h-REA/hREA/blob/8d6010d58f27cf17b8394eac02cc3059afa01729/lib/hdk_time_indexing/src/reading.rs#L15

All the failing tests relate to this, by the looks of it.

The current status of tests is:

```
  205 tests
  608 passed
  18 failed

  Failed Tests:   There were 18 failures

    x query for all agents OK
    x query for all agreements OK
    x all events correctly retrievable
    x event IDs OK
    x event cursors OK
    x all resources correctly retrievable
    x resource IDs OK
    x resource cursors OK
    x query for all fulfillments OK
    x query for all plans OK
    x query for all processes OK
    x query for all commitments OK
    x query for all intents OK
    x query for all proposals OK
    x query for all satisfactions OK
    x query for all process specifications OK
    x query for all resource specifications OK
    x query for all units OK
```

So far, running this:
```
cd test
TRYORAMA_LOG_LEVEL=debug WASM_LOG=debug RUST_LOG=\"debug,wasmer_compiler_cranelift=error,holochain::core::workflow=error,holochain::conductor::entry_def_store=error,holochain::core::ribosome::host_fn::get_links=error,holochain::conductor::api::api_external::app_interface=error,holochain::core::ribosome::host_fn::get=error,holochain::conductor::conductor::network_impls=error,holochain::core::queue_consumer::integrate_dht_ops_consumer=error\" RUST_BACKTRACE=1 GRAPHQL_DEBUG=1 tape agreement/test_agreement_crud.js
```

And with some special logging output, I see this:
```
[GraphQL]:  Query.agreements
OUTPUT AGREEMENTS [{"file":"lib/hdk_records/src/lib.rs","line":107,"error":{"Guest":"No index found at address uhCEk_uFAF_fSQYPee_rggWvJYSBP5r-wqE98TpTNMiGOMuR8Ho1t"}},{"file":"lib/hdk_records/src/lib.rs","line":107,"error":{"Guest":"No index found at address uhCEk_uFAF_fSQYPee_rggWvJYSBP5r-wqE98TpTNMiGOMuR8Ho1t"}},{"file":"lib/hdk_records/src/lib.rs","line":107,"error":{"Guest":"No index found at address uhCEk_uFAF_fSQYPee_rggWvJYSBP5r-wqE98TpTNMiGOMuR8Ho1t"}},{"file":"lib/hdk_records/src/lib.rs","line":107,"error":{"Guest":"No index found at address uhCEkQBLPCmlVwRpEuLshng1hS6ofmhkAr75hYEM2cpPZXdh2OvBb"}},{"file":"lib/hdk_records/src/lib.rs","line":107,"error":{"Guest":"No index found at address uhCEk70X16EztcDKKta8l_X9GBIvQo45E0fN54vQXkDw5DRVHoFzb"}},{"file":"lib/hdk_records/src/lib.rs","line":107,"error":{"Guest":"No index found at address uhCEku67-MXnUgMaie15ZiKdTT3tiYpTCcCfBPeJdvDNYzMv1h2ql"}}]
[result]:  {
  pageInfo: {
    startCursor: 'uhCEkMLNlNpGkHodhlylz3d3IrME9R35NCCHyDk_O84uyVt1gZq7J:uhC0kpu5lntBBQyecpdvJaM_dMcLypWu3c6exIcjbPMEZIFEdS7vM',
    endCursor: 'uhCEkDOS84if_7CC5kVvFCZzZNGxmK1AOMC63ZVaYiC89FSt5BB4t:uhC0kpu5lntBBQyecpdvJaM_dMcLypWu3c6exIcjbPMEZIFEdS7vM',
    hasPreviousPage: true,
    hasNextPage: true
  },
  edges: [
    {
      node: [Object],
      cursor: 'uhCEkMLNlNpGkHodhlylz3d3IrME9R35NCCHyDk_O84uyVt1gZq7J:uhC0kpu5lntBBQyecpdvJaM_dMcLypWu3c6exIcjbPMEZIFEdS7vM'
    },
    {
      node: [Object],
      cursor: 'uhCEkDOS84if_7CC5kVvFCZzZNGxmK1AOMC63ZVaYiC89FSt5BB4t:uhC0kpu5lntBBQyecpdvJaM_dMcLypWu3c6exIcjbPMEZIFEdS7vM'
    },
    {
      node: [Object],
      cursor: 'uhCEkMLNlNpGkHodhlylz3d3IrME9R35NCCHyDk_O84uyVt1gZq7J:uhC0kpu5lntBBQyecpdvJaM_dMcLypWu3c6exIcjbPMEZIFEdS7vM'
    },
    {
      node: [Object],
      cursor: 'uhCEkDOS84if_7CC5kVvFCZzZNGxmK1AOMC63ZVaYiC89FSt5BB4t:uhC0kpu5lntBBQyecpdvJaM_dMcLypWu3c6exIcjbPMEZIFEdS7vM'
    },
    {
      node: [Object],
      cursor: 'uhCEkMLNlNpGkHodhlylz3d3IrME9R35NCCHyDk_O84uyVt1gZq7J:uhC0kpu5lntBBQyecpdvJaM_dMcLypWu3c6exIcjbPMEZIFEdS7vM'
    },
    {
      node: [Object],
      cursor: 'uhCEkDOS84if_7CC5kVvFCZzZNGxmK1AOMC63ZVaYiC89FSt5BB4t:uhC0kpu5lntBBQyecpdvJaM_dMcLypWu3c6exIcjbPMEZIFEdS7vM'
    }
  ],
  errors: [
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] }
  ]
}
not ok 3 query for all agreements OK
```

Commit:
nix files

Commit:
dna-auth-resolver bumps, without compile changes

Commit:
fix comma spacing

Commit:
update happ manifest format -> role_id to role_name

Commit:
update holochain crates, no compile yet

Commit:
compiling

Commit:
necessary changes for performing authorization of zome calls

pospi commented:
Any ideas about this issue with the Nix update?

```
copying path '/nix/store/67jrknz2a3y1wy14hgr7icmc95iba3w1-rust-overlay-03ec5d9' from 'https://holochain-ci.cachix.org'...
unknown compression method 'zstd'
error: build of '/nix/store/sqrvi964bg2m8bkfzbbpc0m37vb0cc9y-rust-overlay-03ec5d9.drv' failed
```

pospi committed:
update DNA Auth Resolver refs to point to tag

pospi committed:
document Nix version gotcha

pospi committed:
heading update in developer README

pospi commented:
Got the Nix config working- they added zstd support in v2.4 and I was slightly behind that- updating fixed it. Added it as a known gotcha in the developer docs.

pospi committed:
update lockfile

pospi committed:
update lockfile

pospi committed:
update Chrono library usage to avoid deprecated method

pospi commented:
A little more progress made. Recommend you run `npm run clean` and reinstall before proceeding as I was getting some errors with Chrono that you didn't seem to be hitting (fixed in adb9bf7d).

Looks like I am getting the same test errors though, so that's good (:

One of the next jobs would be getting the GraphQL explorer working again, since it uses different init scripts to the tests. If you've already gotten up to date with the latest changes to `hc` would you be able to update `run-dev-conductor.sh`? It's making you enter a passphrase to init the keystore (ideally we can skip this), some command names seem to have changed, and it will be the place we manually specify the conductor admin port so that the matching env var can be declared in the app's NPM package scripts.

pospi commented:
Hm, this is going to be another wild debugging adventure.

Printing out `entries_result` in `#exposed_read_api_method_name` at the macro layer and I'm seeing some strange results.

Sometimes there are 9 values returned in this recurring pattern:
![image](https://user-images.githubusercontent.com/437540/210225559-c6664275-d5c1-4d89-b879-22f074736e21.png)

Other times there are 23 returned values in this recurring pattern:
![image](https://user-images.githubusercontent.com/437540/210225626-ba4ca294-34e4-4a81-ac46-548b4c91c7bc.png)

In all cases the last 3 `IndexNotFound` errors are looking for the same hashes. This is leading me to believe something has gone wrong with the indexing structure and it no longer knows when to stop recursing. The patterns are likely due to rounding and which buckets the actual entries fall into. But it's also pretty weird that the records are being successfully returned in among the list and that in some cases 1 invalid hash is being requested repeatedly and in others, 2 invalid hashes are requested but less frequently.

Anyway, nothing to conclude yet but hopefully this information is helpful.

pospi commented:
Hard to say whether the error is coming from the read or write phase, but `hdk_semantic_indexes_zome_lib::read_remote_entry_identity` is where the errors get thrown. This gets called by `retrieve_foreign_record` in `query_time_index`, which seems to be getting an invalid list from `hdk_semantic_indexes_zome_lib::reading::read_all_entry_hashes` that includes every node in the tree, rather than just the childmost nodes.

No idea why. Did the behaviour of `hdk::get_links` change in some way? Or possibly the init code that determines `INDEX_DEPTH` and `HAS_CHUNK_LEAVES` no longer operates as expected in the new hdk?

Connoropolous commented:
Rad! Thanks for jumping in. 
I had done some similar debugging and landed on similar questions, so we seem to have the same sense of it.

These are good questions, and I'm equally ambiguous. 
"No idea why. Did the behaviour of hdk::get_links change in some way? Or possibly the init code that determines INDEX_DEPTH and HAS_CHUNK_LEAVES no longer operates as expected in the new hdk?"

Less likely to me: Did the behaviour of hdk::get_links change in some way.
I have reviewed the changelogs fairly thoroughly, and unless something got merged without a changelog item, which is possible but not likely, I can't see this being the case.

More likely to me: the init code that determines INDEX_DEPTH and HAS_CHUNK_LEAVES no longer operates as expected in the new hdk
We could test by printing out debug values that this code is init-ing properly and that the values used to drive the recursion aren't weird in any way.

Connoropolous commented:
> One of the next jobs would be getting the GraphQL explorer working again, since it uses different init scripts to the tests. If you've already gotten up to date with the latest changes to `hc` would you be able to update `run-dev-conductor.sh`? It's making you enter a passphrase to init the keystore (ideally we can skip this), some command names seem to have changed, and it will be the place we manually specify the conductor admin port so that the matching env var can be declared in the app's NPM package scripts.

Yes, I can do that bit

pospi commented:
Damn, the globals look good. Something else going on then

```
chunk interval 30s, idx depth [Minute, Hour, Day, Month, Year], has leaves true
```

Connoropolous commented:
I have the feeling it’d be helpful to see the source chain. I think if I got the graphql explorer up, then I might also be able to use command line tools to log out source chain values before and after discrete actions like the creation of an Agreement.

Connoropolous commented:
Or the playground

pospi commented:
Um OK so today (with a fresh tmpdir)

```
$ TRYORAMA_LOG_LEVEL=debug WASM_LOG=debug RUST_LOG="debug,wasmer_compiler_cranelift=error,holochain::core::workflow=error,holochain::conductor::entry_def_store=error,holochain::core::ribosome::host_fn::get_links=error,holochain::conductor::api::api_external::app_interface=error,holochain::core::ribosome::host_fn::get=error,holochain::conductor::conductor::network_impls=error,holochain::core::queue_consumer::integrate_dht_ops_consumer=error" RUST_BACKTRACE=1 GRAPHQL_DEBUG=1 npx tape agreement/test_agreement_crud.js

    [...truncated...]

1..10
# tests 10
# pass  10
```

:thinking: 

...then, re-running:

```
    [...truncated...]

not ok 3 query for all agreements OK
  ---
    operator: equal
    expected: 2
    actual:   10
    stack: |-
      Error: query for all agreements OK
      
    [...truncated...]

1..10
# tests 10
# pass  9
# fail  1
```

:thinking: :thinking:  this isn't another "dirty state persists on the DHT between sandbox runs" thing, is it?

Connoropolous commented:
!!

I tried to eliminate that possibility, but perhaps still haven’t succeeded in getting a unique networkSeed to alter the DnaHash per test run

that would explain why there’s a different number each time

Commit:
fix for api change to run-dev-conductor.sh

Commit:
this, however, admin websocket needs to be made constant

Commit:
Update test.yml

Commit:
Update test.yml

Connoropolous commented:
I can confirm that the DnaHash is not changing per test run, so that's a problem

Commit:
fix so that dna is unique each time, thanks to network_seed taking effect

Connoropolous commented on test/init.js:
```diff
@@ -124,8 +124,16 @@ const buildPlayer = async (dnasToInstall, graphQLAPIOptions) => {
           manifest_version: '1',
           roles: dnasToInstall.map((name) => ({
             name: `hrea_${name}_1`,
+            // https://docs.rs/holochain_types/0.1.0-beta-rc.1/src/holochain_types/app/app_manifest/app_manifest_v1.rs.html#165-180
```
adding this fixed the uniqueness of the dna, because of complying with this code:
https://github.com/holochain/holochain/blob/a28eff8dbd688f59e2d16347185040336a0177f0/crates/holochain_types/src/app/app_manifest/app_manifest_v1.rs#L196-L202

tricksy.

Connoropolous commented:
@pospi ok so I am also seeing it intermittently pass, and intermittently fail. randomly. I cannot see a pattern. 

However, I did realize this: even if the tests pass, this still shows up in around the query for all:

```
 Query.agreements
[result]:  {
  pageInfo: {
    startCursor: 'uhCEkNLWyUeHto43kr_Hv1TMOXacSqOhED3VV3UgC9D11mdQqpAZZ:uhC0kAJJP3N4UEtmtjdvH5uzmGhXg8eqypm76j0M3ZB86FEJsKzai',
    endCursor: 'uhCEkTotHfKm8X5QHdODaBzLZ3WXnXqz1Tuwu7Lxv-ELfwAz2mfPW:uhC0kAJJP3N4UEtmtjdvH5uzmGhXg8eqypm76j0M3ZB86FEJsKzai',
    hasPreviousPage: true,
    hasNextPage: true
  },
  edges: [
    {
      node: [Object],
      cursor: 'uhCEkNLWyUeHto43kr_Hv1TMOXacSqOhED3VV3UgC9D11mdQqpAZZ:uhC0kAJJP3N4UEtmtjdvH5uzmGhXg8eqypm76j0M3ZB86FEJsKzai'
    },
    {
      node: [Object],
      cursor: 'uhCEkTotHfKm8X5QHdODaBzLZ3WXnXqz1Tuwu7Lxv-ELfwAz2mfPW:uhC0kAJJP3N4UEtmtjdvH5uzmGhXg8eqypm76j0M3ZB86FEJsKzai'
    }
  ],
  errors: [
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] },
    { file: 'lib/hdk_records/src/lib.rs', line: 107, error: [Object] }
  ]
}
```

Connoropolous commented:
Re. "dirty state persists on the DHT between sandbox runs"

I have solved the issue (pushed the code) of the DnaHash not changing between test runs, and the issues persist

Commit:
Update test.yml

Connoropolous commented on apps/hrea-graphql-explorer/package.json:
```diff
@@ -24,7 +24,7 @@
   },
   "scripts": {
-    "start": "REACT_APP_HC_APP_ID='hrea_suite' REACT_APP_HC_CONN_URL='ws://localhost:4000' BROWSER=none react-scripts start",
+    "start": "REACT_APP_HC_APP_ID='hrea_suite' REACT_APP_HC_CONN_URL='ws://localhost:4000' REACT_APP_HC_ADMIN_CONN_URL='ws://localhost:55355' BROWSER=none react-scripts start",
```
I can't see in `run-dev-conductor.sh` that I am going to be able to make the admin websocket the same every time. It doesn't seem to be an option. It randomly picks one.

Connoropolous replied:
@pospi I did get the GraphQL explorer working again, but you have edit and set this manually for now, based on whatever `npm run dht:conductor` spits out

pospi commented:
> However, I did realize this: even if the tests pass, this still shows up in around the query for all

I was going to report that yesterday but got pulled away. I think the recursion through the tree has been flaky this whole time, it's returning every child hash rather than just the leaves. We should still fix it, but I think it might not actually be related to these failures. Not getting a clean DHT per test seems like it's the more likely problem at this stage?

Connoropolous commented:
Ah. 
Well, I have resolved the clean DHt thing in recent commits and the issue persists.

pospi commented:
lol, shit, it's January's fault:

```
11:37:22 [Tryorama - Local Conductor] debug: Jan 09 11:37:22.307 DEBUG wasm_trace: hdk_time_indexing::writing:lib/hdk_time_indexing/src/writing.rs:46 WRITE SEgments for index "agreement.created" [IndexSegment(1672531200000), IndexSegment(1672531200000), IndexSegment(1673222400000), IndexSegment(1673226000000), IndexSegment(1673228220000), IndexSegment(1673228220000)]

11:37:22 [Tryorama - Local Conductor] debug: Jan 09 11:37:22.308 DEBUG wasm_trace: hdk_time_indexing::writing:lib/hdk_time_indexing/src/writing.rs:52 EntryHash(uhCEkBjUuNWIl21P37x8IdCi42Xr_c1MqTF9b6zadfnjEX2nHaqG_) ---> 2023-01-01T00:00:00Z
11:37:22 [Tryorama - Local Conductor] debug: Jan 09 11:37:22.309 DEBUG wasm_trace: hdk_time_indexing::writing:lib/hdk_time_indexing/src/writing.rs:66 EntryHash(uhCEku67-MXnUgMaie15ZiKdTT3tiYpTCcCfBPeJdvDNYzMv1h2ql) ---> 2023-01-01T00:00:00Z
11:37:22 [Tryorama - Local Conductor] debug: Jan 09 11:37:22.310 DEBUG wasm_trace: hdk_time_indexing::writing:lib/hdk_time_indexing/src/writing.rs:66 EntryHash(uhCEku67-MXnUgMaie15ZiKdTT3tiYpTCcCfBPeJdvDNYzMv1h2ql) ---> 2023-01-09T00:00:00Z
11:37:22 [Tryorama - Local Conductor] debug: Jan 09 11:37:22.311 DEBUG wasm_trace: hdk_time_indexing::writing:lib/hdk_time_indexing/src/writing.rs:66 EntryHash(uhCEkW1TZCzZ5LiSJM3iLnoOi9_ncelKz5shzQVAO3HOO3oXBfqpS) ---> 2023-01-09T01:00:00Z
11:37:22 [Tryorama - Local Conductor] debug: Jan 09 11:37:22.312 DEBUG wasm_trace: hdk_time_indexing::writing:lib/hdk_time_indexing/src/writing.rs:66 EntryHash(uhCEk-Xnb_wPXLk6aF7eIGN7xG-AiFgsiwKtnsaAq4TsJfWQdSDv5) ---> 2023-01-09T01:37:00Z
11:37:22 [Tryorama - Local Conductor] debug: Jan 09 11:37:22.313 DEBUG wasm_trace: hdk_time_indexing::writing:lib/hdk_time_indexing/src/writing.rs:66 EntryHash(uhCEkTvKvAnuQPgXeUZTGXNuEeFFLJCeeqkalb8lFtc1Ip34oDoNq) ---> 2023-01-09T01:37:00Z
```

Because I am just truncating the timestamp at an appropriate time point to start each period, it ends up looping where it shouldn't if multiple of the timestamps in the tree are the same (a self-referential link is created). This is the case for all of the month of January where the year and month nodes will end up as `2023-01-01T00:00:00Z`.

Sooo I think the solution will be to not attempt to have such a fancy efficient storage format for them, or to place null bytes in there where the truncation starts. Probably the former- I'll just revert to strings to keep it simple since this kind of efficiency is no longer an issue since baseless links became an option.

Connoropolous commented:
> it's January's fault

Of course it is 🤦 cause writing code to handle calendaring

Connoropolous commented:
are you at it?

pospi commented:
Yeah. It's doing my head in a bit but I am most of the way through string conversions I think.

pospi committed:
update internals of time indexing DHT structure to use uniquely truncated strings for each node in the tree

this prevents circular references for overlapping period starts (eg. 2020-01-01 is the start of both '2020' and 'Jan 2020' AND 'Jan 1st 2020') resulting in broken & duplicate entries being returned to the caller

pospi committed:
remove unnecessary entrydef for IndexSegment

pospi committed:
use chrono format() for initialising time segment strings to obtain correct leading 0s in time index segment nodes

pospi committed:
remove unnecessary code in time index segment -> DateTime conversion method

pospi committed:
refix f7d683a4 to ensure that decoding takes into account index segments that have been truncated

pospi commented:
Yay!! happy to merge this if you are :smile:

Connoropolous commented:
You beat the final boss!

Yeah right on, very happy to merge

Connoropolous commented:
nix isn't working on MacOS for now, that's on holochain todo list.
